[NNC] Make splitWithMask transform in-place

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58270 [NNC] Add documentation for splitWith APIs
* **#58269 [NNC] Make splitWithMask transform in-place**
* #58268 [NNC] Make splitWithTail transform in-place

Differential Revision: [D28427227](https://our.internmc.facebook.com/intern/diff/D28427227)

navahgar committed:
[NNC] Make splitWithMask transform in-place

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427227).

navahgar committed:
Update on "[NNC] Make splitWithMask transform in-place"


Differential Revision: [D28427227](https://our.internmc.facebook.com/intern/diff/D28427227)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427227).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1250,10 +1250,11 @@ void LoopNest::splitWithMask(For* f, int factor, For** outer, For** inner) {
-      new For(i_outer, new IntImm(0), split_count, *inner, f->loop_options());
-
-  p->replace_stmt(f, *outer);
+  f->setVar(i_outer);
```
We' re still cloning `body_inner` on line 1236. Can we avoid that? Otherwise, these loops would be invalidated:

Before Split:
```
for i in 0..N:
  for j in 0..100:
   ...
```

After Split:
```
for io in 0..8:
  for ii in 0..N/8:
    for j_new in 0..100: # << this is not the same loop as loop 'j'
     ...
```

Can we also add a test for such case?

navahgar replied:
Good catch with the cloning of `body_inner`. I removed that in `splitWithTail` but missed it here. Fixed.

However, that is not enough to ensure that the body remains the same. We need to fix the `IRMutator` APIs as well as `Substitute` for that. I'm starting to work on that. Once that is fixed, I will add a test here to check the same.

navahgar committed:
Update on "[NNC] Make splitWithMask transform in-place"


Differential Revision: [D28427227](https://our.internmc.facebook.com/intern/diff/D28427227)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427227).

ZolotukhinM commented on benchmarks/cpp/tensorexpr/bench_gemm.cpp:
```diff
@@ -81,16 +81,14 @@ BENCHMARK_DEFINE_F(Gemm, TensorExprTile32x32)(benchmark::State& state) {
-    te::For* mo;
     te::For* mi;
-    loop.splitWithMask(m, 32, &mo, &mi);
+    loop.splitWithMask(m, 32, &mi);
```
Nit: we could remove `mi` here entirely since it's not used (the same in some other places).

navahgar replied:
Done.

ZolotukhinM reviewed (approved):
👍

navahgar committed:
Update on "[NNC] Make splitWithMask transform in-place"


Differential Revision: [D28427227](https://our.internmc.facebook.com/intern/diff/D28427227)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427227).

navahgar committed:
Update on "[NNC] Make splitWithMask transform in-place"


Differential Revision: [D28427227](https://our.internmc.facebook.com/intern/diff/D28427227)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427227).

navahgar committed:
Update on "[NNC] Make splitWithMask transform in-place"


Differential Revision: [D28427227](https://our.internmc.facebook.com/intern/diff/D28427227)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427227).
